sdk/lib/rtl/heap.c

```diff
@@ -188,9 +215,13 @@ RtlpInitializeHeap(OUT PHEAP Heap,
     for (Index = 0; Index < HEAP_SEGMENTS; ++Index)
         Heap->Segments[Index] = NULL;
 
-    /* Initialise the Heap Free Heap Entry lists */
-    for (Index = 0; Index < HEAP_FREELISTS; ++Index)
-        InitializeListHead(&Heap->FreeLists[Index]);
+    /* Initialise the free entry lists. */
+    InitializeListHead(&Heap->FreeLists);
+    RtlInitializeBitMap(&Heap->FreeHintBitmap,
+                        (PULONG)&Heap->FreeHints[DeCommitFreeBlockThreshold],
+                        DeCommitFreeBlockThreshold);
+    RtlClearAllBits(&Heap->FreeHintBitmap);
+    RtlZeroMemory(&Heap->FreeHints[0], sizeof(Heap->FreeHints[0]) * DeCommitFreeBlockThreshold);
```

Comment:
This code feels insecure when `DeCommitFreeBlockThreshold` happens to be equal to zero.

(This is possible by manually setting one's own `Peb->HeapDeCommitFreeBlockThreshold` value to zero, and then creating a new heap with `RtlCreateHeap()` with also `Parameters->DeCommitFreeBlockThreshold` set to zero.)

Reply:
Now DeCommitFreeBlockThreshold is forced to be PAGE_SIZE (which is the default value). Bigger size will be unsupported anyway, as that would make the heap header bigger than PAGE_SIZE.